pkg/apis/network/v1/egressrouter_types.go

```diff
@@ -0,0 +1,275 @@
+	// Arguments specific to the interfaceType macvlan
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default:={mode: Bridge}
+	Macvlan MacvlanConfig `json:"macvlan"`
+}
+
+// MacvlanMode defines the different types of modes that are supported for the macvlan interface.
+// source: https://man7.org/linux/man-pages/man8/ip-link.8.html
+type MacvlanMode string
+
+const (
+	// MacvlanModeBridge connects all endpoints directly to each other, communication is not redirected through the physical interface's peer.
+	MacvlanModeBridge MacvlanMode = "Bridge"
+
+	// MacvlanModePrivate does not allow communication between macvlan instances on the same physical interface,
+	// even if the external switch supports hairpin mode.
+	MacvlanModePrivate MacvlanMode = "Private"
+
+	// MacvlanModeVEPA is the Virtual Ethernet Port Aggregator mode. Data from one macvlan instance to the other on the
+	// same physical interface is transmitted over the physical interface. Either the attached switch needs
+	// to support hairpin mode, or there must be a TCP/IP router forwarding the packets in order to allow
+	// communication. This is the default mode.
+	MacvlanModeVEPA MacvlanMode = "VEPA"
+
+	// MacvlanModePassthru mode gives more power to a single endpoint, usually in macvtap mode.
+	// It is not allowed for more than one endpoint on the same physical interface. All traffic will be forwarded
+	// to this endpoint, allowing virtio guests to change MAC address or set promiscuous mode in order to bridge the
+	// interface or create vlan interfaces on top of it.
+	MacvlanModePassthru MacvlanMode = "Passthru"
+)
+
+// MacvlanConfig consists of arguments specific to the macvlan EgressRouterInterfaceType
+type MacvlanConfig struct {
+	// Mode depicts the mode that is used for the macvlan interface; one of Bridge|Private|VEPA|Passthru. The default mode is "Bridge".
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum="Bridge";"Private";"VEPA";"Passthru"
+	// +kubebuilder:default:="Bridge"
+	Mode MacvlanMode `json:"mode"`
+
+	// Name of the master interface. Need not be specified if it can be inferred from the IP address.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:Optional
+	// +optional
+	Master string `json:"master,omitempty"`
+}
+
+// ProtocolType defines the protocol types that are supported
+type ProtocolType string
+
+const (
+	// ProtocolTypeTCP refers to the TCP protocol
+	ProtocolTypeTCP ProtocolType = "TCP"
+
+	// ProtocolTypeUDP refers to the UDP protocol
+	ProtocolTypeUDP ProtocolType = "UDP"
+
+	// ProtocolTypeSCTP refers to the SCTP protocol
+	ProtocolTypeSCTP ProtocolType = "SCTP"
+)
+
+// EgressRouterModeConfig represents a single mode's configuration.
+// Mode must be specified along with exactly one "Config" that matches the mode.
+// Each config consists of parameters used to define the remote destination that the router pod is allowed to connect to for that mode.
+type EgressRouterModeConfig struct {
```

Comment:
You don't need this struct if you don't want. Look at the way Volumes are configured: https://v1-18.docs.kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#volume-v1-core

I would just do something like

```go
type EgressRouterSpec struct {
    Mode EgressRouterMode
    Redirect *RedirectConfig
}
type RedirectConfig struct {
    AllowedDestinations []IPDestination
}
etc
```

Reply:
yea, the reason why I split it out was to make it more modular and then use the `minProperties` and `maxProperties` on it. Since `networkInterface` and `addresses` fields are required fields, I guess we could still validate if one of the config's is provided.

Reply:
https://etherpad.opendev.org/p/CubpCqAkGAadpbruAF5K

Reply:
@danwinship : so casey came up with this idea (etherpad), where we could just combine the single IP destination case on 3.11 with fallbackIP idea for the multiple destinations into a single field called `FallBackIP` and then put the rest into a separate struct and make the `ip, port, protocol` parts `required`. Let me know in case you have any suggestions/changes regarding this approach. We sort of need to get this CRD in asap.

The only difference between this approach and previous one I had was that now we can't accept a format that looks like:
```
destinations: [ipA, ipB, ipC]
```
but then that use case is not supported in 3.11 so should be fine.

Reply:
If there's a use case that this config doesn't support, then we should certainly try and support it. I'm not sure what an array if IPs would mean though - that traffic would be spread across them on a per-connection basis?

Reply:
> If there's a use case that this config doesn't support, then we should certainly try and support it. I'm not sure what an array of IPs would mean though - that traffic would be spread across them on a per-connection basis?

Yea I guess? although it doesn't make much sense, because we don't do any control on connection basis on the router end. Based on IP table rule matches it might be sent to both destinations right? I had the same-ish (a bit different) question: https://github.com/openshift/cluster-network-operator/pull/852#discussion_r577546026 on what would happen and if we have such a use case or not.

As far as I can see in 3.11 we don't support this use case, there can only be one line with a single IP (https://github.com/openshift/images/blob/master/egress/router/egress-router.sh#L70-L109)

The reason I had this doubt was because of the lines here: https://github.com/openshift/egress-router-cni/blob/master/README.md#network-configuration-reference in the `destination` field part.

I think we can safely discard this use case unless there is a strong reason that the CU might want to do something like that.

Reply:
agreed :+1: